### PR TITLE
Tactic to automate some routine set theory by reduction to FOL

### PR DESCRIPTION
**Pull request by binghe**
Hi,

the main purpose of this PR is to move the following 3 new tools from various scripts (in `probability`) into `pred_setLib`. They're "tactics to automate some routine set theory by reduction to FOL" (the original comments), originally ported from HOL Light ([`sets.ml`](https://github.com/jrh13/hol-light/blob/master/sets.ml)) by people at Concordia HVG.

```
fun SET_TAC L =
    POP_ASSUM_LIST (K ALL_TAC) \\
    rpt COND_CASES_TAC \\
    REWRITE_TAC (append [EXTENSION, SUBSET_DEF, PSUBSET_DEF, DISJOINT_DEF,
                         SING_DEF] L) \\
    SIMP_TAC std_ss [NOT_IN_EMPTY, IN_UNIV, IN_UNION, IN_INTER, IN_DIFF,
      IN_INSERT, IN_DELETE, IN_REST, IN_BIGINTER, IN_BIGUNION, IN_IMAGE,
      GSPECIFICATION, IN_DEF, EXISTS_PROD] \\
    METIS_TAC [];

fun ASM_SET_TAC L = rpt (POP_ASSUM MP_TAC) >> SET_TAC L;

fun SET_CONV tm = prove (tm, SET_TAC []);
```

The `SET_CONV` was named `SET_RULE` (also in HOL Light), I changed its name to `SET_CONV` as its type actually fit the type of other conversions.

They're very useful. With `SET_TAC`, many small set theorems now have trivial proofs, e.g.

```
Theorem DISJOINT_RESTRICT_L :
  !s t c. DISJOINT s t ==> DISJOINT (s INTER c) (t INTER c)
Proof SET_TAC []
QED
```

Once they're moved into `pred_setLib`, many proof scripts in `src/probability` now need to depend on `pred_setLib` with their own definitions of `SET_TAC` etc. removed. I also made some other small changes, e.g. fixing some `pred_set` theorems with missing quantifiers added. In summary, all these changes were from my ongoing probability work, they're all needed. (But if you think any of them is not general enough, I can remove it and move it back to the places where they're just needed.) Below is the full description of these other changes:

## Changes in `pred_setTheory`

Beside fixing the missing quantifiers of several theorems (`DISJOINT_DIFFS`, `INJ_DELETE`, `INJ_EXTEND`, `INJ_IMAGE`, `INJ_IMAGE_SUBSET`, `COUNT_11`, `SUBSET_BIGUNION_I`), I added several new small theorems (to enrich the related definitions):

```
   [SUBSET_DISJOINT]  Theorem      
      ⊢ ∀s t u v. DISJOINT s t ∧ u ⊆ s ∧ v ⊆ t ⇒ DISJOINT u v

   [IMAGE_SING]  Theorem      
      ⊢ ∀f x. IMAGE f {x} = {f x}

   [COUNT_DELETE]  Theorem
      ⊢ ∀n. count n DELETE n = count n
   
   [COUNT_MONO]  Theorem
      ⊢ ∀m n. m ≤ n ⇒ count m ⊆ count n
   
   [COUNT_NOT_EMPTY]  Theorem
      ⊢ ∀n. 0 < n ⇔ count n ≠ ∅

   [EMPTY_IN_POW]  Theorem
      ⊢ ∀s. ∅ ∈ POW s

   [DIFF_INTER_COMPL]  Theorem      
      ⊢ ∀s t. s DIFF t = s ∩ COMPL t
```

## Changes in `realTheory` and `real_sigmaTheory`

In `realTheory`, I added the following simple versions of Archimedes axioms (c.f. `REAL_ARCH`) for real numbers:

```
   [SIMP_REAL_ARCH]  Theorem
      ⊢ ∀x. ∃n. x ≤ &n
   
   [SIMP_REAL_ARCH_NEG]  Theorem
      ⊢ ∀x. ∃n. -&n ≤ x

   [REAL_ARCH_POW]  Theorem      
      ⊢ ∀x. ∃n. x < 2 pow n

   [REAL_LT_RNEG]  Theorem      
      ⊢ ∀x y. x < -y ⇔ x + y < 0
```

The last one, `REAL_LT_RNEG`, is a dual version of `REAL_LE_RNEG`.

In `real_sigmaTheory` two more theorems about the sum image of constant elements were found in `examples/diningcryptos` and moved here:

```
   [REAL_SUM_IMAGE_FINITE_CONST2]  Theorem
      ⊢ ∀P.
            FINITE P ⇒
            ∀f x. (∀y. y ∈ P ⇒ f y = x) ⇒ SIGMA f P = &CARD P * x
   
   [REAL_SUM_IMAGE_FINITE_CONST3]  Theorem      
      ⊢ ∀P. FINITE P ⇒ ∀c. SIGMA (λx. c) P = &CARD P * c
```

Above `REAL_SUM_IMAGE_FINITE_CONST2` is more general/useful than the original `REAL_SUM_IMAGE_FINITE_CONST` which requires the function `f` being defined in a full domain:

```
   [REAL_SUM_IMAGE_FINITE_CONST]  Theorem      
      ⊢ ∀P. FINITE P ⇒ ∀f x. (∀y. f y = x) ⇒ SIGMA f P = &CARD P * x
```

## Changes in `src/probability`

Beside adding `pred_setLib` into dependencies, I moved the definition "from_def" and some related theorem from `real_topologyTheory` to `util_probTheory` and further enriched them, below is the list of these theorems:

```
   [from_def]  Definition      
      ⊢ ∀n. from n = {m | n ≤ m}

   [FROM_0]  Theorem      
      ⊢ from 0 = 𝕌(:num)

   [IN_FROM]  Theorem      
      ⊢ ∀m n. m ∈ from n ⇔ n ≤ m

   [DISJOINT_COUNT_FROM]  Theorem
      ⊢ ∀n. DISJOINT (count n) (from n)
   
   [DISJOINT_FROM_COUNT]  Theorem      
      ⊢ ∀n. DISJOINT (from n) (count n)

   [UNION_COUNT_FROM]  Theorem     
      ⊢ ∀n. count n ∪ from n = 𝕌(:num)
   
   [UNION_FROM_COUNT]  Theorem      
      ⊢ ∀n. from n ∪ count n = 𝕌(:num)
```

## Changes in `arithmeticTheory`

In `arithmeticTheory` there's a definition of `DIV2` but no supporting theorem at all:

```
   [DIV2_def]  Definition      
      ⊢ ∀n. DIV2 n = n DIV 2
```

I found it useful in probability applications, and I needed a theorem to (automatically) simplify `DIV2 (2 * n)` back to `n`. Now I moved this theorem here:

```
   [DIV2_DOUBLE]  Theorem      
      ⊢ ∀n. DIV2 (2 * n) = n
```

Besides, I found the existing TeX notation definition for the "minus" operator actually doesn't work, maybe because the notation is defined on ASCII char "-" but its unicode version. Any way, I added another one for the unicode version and it solved my problem:

```
val _ = TeX_notation {hol = UTF8.chr 0x2212, TeX = ("\\ensuremath{-}", 1)}
```

## Changes in examples and final words

I added a `Holmakefile` at `examples/miller` (toplevel), now there's no need to put its sub-folders separately into the build sequence (`tools/sequences/final-examples`).

Basically those changes (in `pred_set`, `real`, etc.) are all I needed to build my ongoing [probability work](https://github.com/binghe/HOL-Probability) outside of HOL. I shall continue working on it (probability) and eventually submit it to HOL official.

### CODE REVIEW

**Comment by binghe** [comment]
Sorry, `IMAGE_EMPTY_FUN` is canceled. (The existing `IMAGE_EQ_SING` is more general.)

**Comment by mn200** [comment]
Thanks for this!  It all looks extremely useful and correct. Some minor things:

- Please keep `SET_RULE` as `SET_RULE`; though its type is correct for a conversion, it doesn't  return an equality on the term argument. (This is why, for example, `numLib` has `ARITH_PROVE` and `ARITH_CONV`; the former wraps `EQT_ELIM` around the latter.)
- Could you write `.doc` files for `SET_TAC` and `SET_RULE` in `help/Docfiles` please? (I can do this, if you prefer not to.)
- Please consider moving `SET_TAC` to `bossLib`.

**Comment by binghe** [comment]
Hi Michael,

thanks for your approval and suggestions. Now I have moved all three functions to `bossLib` and added documentations -- both doc files and a paragraph in `next-release.md`.  Please check and let me know if they are not sufficient.

You didn't explicitly mention `ASM_SET_TAC`, but I guess you meant them all, thus eventually I didn't touch `pred_setLib` at all.  Although `METIS_TAC` always take assumptions into account (c.f. `MESON_TAC` v.s. `ASM_MESON_TAC`), the definition of `SET_TAC` actually explicitly removed all assumptions before calling `METIS_TAC`, thus the behavior is different with `ASM_SET_TAC`.  From their uses in `real_topologyTheory` (by John Harrison), I believe both tactics are useful. (Maybe `SET_TAC` sometimes has much better performance under a huge assumption list from the current goal.)

--Chun

**Comment by binghe** [comment]
I have made too many commits in this branch, please consider squashing all commits into just one, when you click the Merge button.

**Comment by mn200** [comment]
Thanks for your work on this!
